src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,37 +118,19 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
-
-            _entryLock.EnterWriteLock();
-            try
+            if (_entries.TryRemove(entry.Key, out priorEntry))
             {
-
-                if (_entries.TryGetValue(entry.Key, out priorEntry))
-                {
-                    _entries.Remove(entry.Key);
-                    priorEntry.SetExpired(EvictionReason.Replaced);
-                }
-
-                if (!entry.CheckExpired(utcNow))
-                {
-                    _entries[entry.Key] = entry;
-                    entry.AttachTokens();
-                    added = true;
-                }
-            }
-            finally
-            {
-                _entryLock.ExitWriteLock();
+                priorEntry.SetExpired(EvictionReason.Replaced);
+                priorEntry.InvokeEvictionCallbacks();
```

Comment:
I thought it wasn't important to keep the order but I suppose it makes sense to avoid behaviour changes.